### PR TITLE
Support for version hashes, for GHC 7.12.

### PR DESCRIPTION
**Pull request by ezyang**
In GHC 7.12, package key generation moves from Cabal to GHC.
Version hashes are the new GHC 7.12 mechanism for letting
Cabal communicate information about the package dependency
tree to GHC.  See:
https://ghc.haskell.org/trac/ghc/ticket/10566
for some background details.  So this patch:

    - Defines version hashes, which are a hash of the source package
      ID and the version hashes of all internal dependencies, and

    - Queries GHC during configuration to calculate a package key.

A major refactoring which is part of this commit, and in fact
takes up more lines than version hashes proper, is moving package
keys from LocalBuildInfo to LibComponentBuildInfo.  If you have
an LBI, you can still extract a package key using the new
localPkgKey function (which looks through the ComponentBuildInfos
of a LocalBuildInfo for the library in question).  This is conceptually
cleaner for two reasons:

    1. Only dependencies of the *library* are counted as part
    of the package key, as opposed to *all* dependencies which
    we previously used.

    2. A package key doesn't really mean much for an executable,
    or a test suite, since no one else will have to link against
    them.  So we can fall back on something simpler.

Bug: abi-hash needs to be passed package key.  I'll amend it to this
PR shortly.

Signed-off-by: Edward Z. Yang <ezyang@cs.stanford.edu>

### CODE REVIEW

**Comment by dcoutts** [comment]
Sounds ok in principle. I've not looked at the changes in detail yet.

Could you fix the build failure?

**Comment by ezyang** [comment]
After lots of scuffling around, the test are finally green!

File: Cabal/Distribution/Package.hs [hunk 1]
```diff
@@ -303,6 +278,51 @@ instance NFData PackageKey where
...
 
 -- ------------------------------------------------------------
+-- * Version hashes
+-- ------------------------------------------------------------
+
+-- | A hash of a package ID and the version hashes of all textual
+-- dependencies, this uniquely identifies the version resolution
+-- of Cabal.  Create using 'mkVersionHash'.  (In fact, this
+-- formulation works even with private dependencies.)
+data VersionHash
```
Comment (23Skidoo) [inline Cabal/Distribution/Package.hs]: Why not a `newtype`?

File: Cabal/Distribution/Simple/LocalBuildInfo.hs [hunk 2]
```diff
@@ -154,6 +152,13 @@ data LocalBuildInfo = LocalBuildInfo {
 
 instance Binary LocalBuildInfo
 
+localPkgKey :: LocalBuildInfo -> PackageKey
```
Comment (23Skidoo) [inline Cabal/Distribution/Simple/LocalBuildInfo.hs]: Please add a Haddock comment.

File: Cabal/Distribution/Package.hs (hunk 1 above)
Comment (ezyang) [inline Cabal/Distribution/Package.hs]: Oops, that's a hold-over from when it had two fields. Will fix.

**Comment by ezyang** [comment]
OK, I've pushed some relatively major changes to the diff; in particular they're no longer called "Version hashes" but "Library names" (subsuming the existing concept in Cabal), and I no longer need to call GHC in order to get a package key, which simplifies matters a great deal.

**Comment by 23Skidoo** [comment]
The new version fails on Travis.

**Comment by 23Skidoo** [comment]
> in particular they're no longer called "Version hashes" but "Library names"

Guess that https://ghc.haskell.org/trac/ghc/wiki/Commentary/Packages/Concepts should also be updated.

File: Cabal/Distribution/Package.hs [hunk 3]
```diff
@@ -303,6 +268,38 @@ instance NFData PackageKey where
     rnf (OldPackageKey pid) = rnf pid
 
 -- ------------------------------------------------------------
+-- * Library names
+-- ------------------------------------------------------------
+
+-- | A library name consists of not only a source package
+-- id ('PackageId') but also the library names of all textual
+-- dependencies; thus, a library name uniquely identifies an
```
Comment (23Skidoo) [inline Cabal/Distribution/Package.hs]: s/an/all/

File: Cabal/Distribution/Package.hs [hunk 4]
```diff
@@ -303,6 +268,38 @@ instance NFData PackageKey where
...
+-- installed packages up to the dependency resolution done by Cabal.
+-- Create using 'packageKeyLibraryName'.  Library names are opaque,
+-- Cabal defined strings.
```
Comment (23Skidoo) [inline Cabal/Distribution/Package.hs]: s/Cabal defined/Cabal-defined/

File: Cabal/Distribution/Simple/Build.hs [hunk 5]
```diff
@@ -406,8 +407,9 @@ testSuiteLibV09AsLibAndExe pkg_descr
     libClbi = LibComponentLocalBuildInfo
                 { componentPackageDeps = componentPackageDeps clbi
                 , componentPackageRenaming = componentPackageRenaming clbi
-                , componentLibraries = [LibraryName (testName test)]
+                , componentLibraryName = LibraryName "test"
```
Comment (23Skidoo) [inline Cabal/Distribution/Simple/Build.hs]: Why doesn't this use the test suite name?

Comment (23Skidoo) [inline Cabal/Distribution/Simple/Build.hs]: Since we can have multiple test suites in the same package, won't this cause problems?

Comment (ezyang) [inline Cabal/Distribution/Simple/Build.hs]: I tried to construct a failing test-case for this, but the detailed-0.9 doesn't publish a test library, so you can't actually use the two test suites together. Still, it costs little to fix. I'll update the V09 to also test multiple libraries.

Comment (ezyang) [inline Cabal/Distribution/Simple/Build.hs]: I guess the generated libraries do clobber. OK, well, pushing fix shortly...

**Pull request merged**